### DISCUSSION TITLE
Add xhtml:link to sitemap.xml

### DISCUSSION THREAD

**Discussion by ctkc** [vercel/next.js#53540] [Help] [answered]
### Goals

- Add "alternate" links in the generated sitemap to specify whether the page is available in other languages or not. This is used by **Google** to better understand the content of the page when it is present in multiple languages

### Non-Goals

-

### Background

This feature is essential for websites in multiple languages. Although we can generate the sitemap manually would we good to also add this new functionality to the NextJs sitemap generator 

### Proposal

According to the Google [documentation](https://developers.google.com/search/docs/specialty/international/localized-versions#example_2) we need to generate a sitemap with the following tag for each language:

```xml
<xhtml:link rel="alternate" hreflang="es" href="https://www.example.com/es/page.html"/>
```

I think we can follow the same approach that is already available in the [Metadata](https://nextjs.org/docs/app/api-reference/functions/generate-metadata#alternates) object so the following code

```javascript
// app/sitemap.xml
export default function sitemap(): MetadataRoute.Sitemap {
  return [
    {
      url: 'https://www.example.com/page.html',
      alternates: {
        languages: {
          es: 'https://www.example.com/es/page.html',
          de: 'https://www.example.com/de/page.html'
        }
      },
      lastModified: new Date(),
    },
  ]
}
```

Should generate the following output

```xml
<urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9" xmlns:xhtml="http://www.w3.org/1999/xhtml">
  <url>
    <loc>https://www.example.com/page.html</loc>
    <xhtml:link
      rel="alternate"
      hreflang="es"
      href="https://www.example.com/es/page.html"/>
    <xhtml:link
      rel="alternate"
      hreflang="de"
      href="https://www.example.com/de/page.html"/>
  </url>
</urlset>
```

In the future we could use the same `alternate` key to include other type of metadata to the sitemap

**Comment by janpaepke** [comment]
Imho this is very sensible and should be added.
@ctkc already supplied a PR for this: https://github.com/vercel/next.js/pull/53765
The implementation is particularly well-crafted as it mirrors the data structure of the Metatags, which makes this very consistent and easy to understand.

**Reply by EdVonTech** [reply]
Is this implemented yet?

**Comment by StepsOnes** [comment]
when will it be added?

**Comment by johannsonic** [comment]
Also need this!

**Comment by codefln** [comment]
This is already described in the documentation, but it doesn't work in the version 14.1.4
https://nextjs.org/docs/app/api-reference/file-conventions/metadata/sitemap#generate-a-localized-sitemap

**Comment by testsnake** [comment]
Looks like it was just implemented https://github.com/vercel/next.js/commit/e4cd547a505c8380cbf010a78f1e2e3ade0f2307

**Comment by IlamathiIla** [comment]
Still the issue is persist, please update this feature ASAP.!

**Reply by testsnake** [reply]
It's still only available in the canary channel. Given it was merged after the release of 14.2.0, I imagine it'll be merged to stable for 14.3.0, whenever that releases.

**Comment by johannsonic** [comment]
For everyone who is looking for a temporary solution: https://www.reddit.com/r/nextjs/comments/161cbms/creating_a_sitemap_with_translated_alternate/

Worked like a charm for me

**Comment by xoocode** [comment]
I must be missing something, but the file in your example does not render correct html, since the link seems to be missing from the http://www.w3.org/1999/xhtml namespace. Changing it to http://www.w3.org/TR/xhtml11/xhtml11_schema.html however works.

Does not render in a brower and is not recognied by google search console:
`<?xml version="1.0" encoding="UTF-8"?>
<urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9" xmlns:xhtml="http://www.w3.org/1999/xhtml">
  <url>
    <loc>https://www.example.com/page.html</loc>
    <xhtml:link
      rel="alternate"
      hreflang="es"
      href="https://www.example.com/es/page.html"/>
    <xhtml:link
      rel="alternate"
      hreflang="de"
      href="https://www.example.com/de/page.html"/>
  </url>
</urlset>`

Renders in a browser and is recognized by google search console:
`<?xml version="1.0" encoding="UTF-8"?>
<urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9" xmlns:xhtml="http://www.w3.org/TR/xhtml11/xhtml11_schema.html">
  <url>
    <loc>https://www.example.com/page.html</loc>
    <xhtml:link
      rel="alternate"
      hreflang="es"
      href="https://www.example.com/es/page.html"/>
    <xhtml:link
      rel="alternate"
      hreflang="de"
      href="https://www.example.com/de/page.html"/>
  </url>
</urlset>`

This creates a problem, since next 14.2 prints out "http://www.w3.org/1999/xhtml" from sitemap.ts.

**Comment by leerob** [comment, marked as answer]
This is available on the Next.js 15 canary release, which will be stable soon.

https://rc.nextjs.org/docs/app/api-reference/file-conventions/metadata/sitemap#generate-a-localized-sitemap

**Reply by dupcer** [reply]
In documentation it references only to App. Is it supported for Page Router?

**Reply by kliatskiidenis** [reply]
> This is available on the Next.js 15 canary release, which will be stable soon.
> 
> https://rc.nextjs.org/docs/app/api-reference/file-conventions/metadata/sitemap#generate-a-localized-sitemap

Can you please tell me if I can get this functionality in the Next 14.1.3 version?”
Because we are afraid to update our project, which is already in production and is expanding its functionality very dynamically, to version 15, lest there are errors that may appear after the transition to the new Next version or there is a need to rewrite the code for the new version because, unfortunately, there is no free time now.

Or maybe it is possible to make a sitemap.xml file statically manually with the ability to specify 
<xhtml:link
      rel="alternate”
      hreflang="es”
      href=“https://acme.com/es”/>

because even if I create a file app/sitemap.xml and copy the code from the documentation into it
<img width="947" alt="Знімок екрана 2025-01-27 о 11 34 09" src="https://github.com/user-attachments/assets/88c3d1d2-fdb9-4670-b179-793d70bc8516" />

then in the browser I don't see the desired sitemap.xml and see this content
<img width="1038" alt="Знімок екрана 2025-01-27 о 11 29 46" src="https://github.com/user-attachments/assets/2485f1fa-789f-49cd-8f4c-2bcc851672af" />

if you remove the generation of <xhtml:link
      rel="alternate”
      hreflang="es”
      href=“https://acme.com/es”/> 
then it is displayed correctly


I tried to create sitemap.xml in the public directory, but it doesn't work as it should

**Reply by nateriverdev** [reply]
you can refer this link:
https://kpmg.com/tr/tr/sitemap.xml

The main point is in this part — just copy it and replace it in your code, and it will work perfectly. Give it a try!
<urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:xhtml="http://www.w3.org/1999/xhtml/" xmlns:image="http://www.google.com/schemas/sitemap-image/1.1" xsi:schemaLocation="http://www.sitemaps.org/schemas/sitemap/0.9 http://www.sitemaps.org/schemas/sitemap/0.9/sitemap.xsd http://www.w3.org/1999/xhtml http://www.w3.org/2002/08/xhtml/xhtml1-strict.xsd">

**Comment by bennajah** [comment]
There is a bug When included alternates
And sitemap render as html document instead xml

Tested in last release v16.0.3
